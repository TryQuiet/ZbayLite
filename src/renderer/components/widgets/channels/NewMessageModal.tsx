import React from 'react'
import { Formik, Form } from 'formik'
import { Typography } from '@material-ui/core'
import Grid from '@material-ui/core/Grid'
import { makeStyles } from '@material-ui/core/styles'
import Button from '@material-ui/core/Button'
import TextField from '@material-ui/core/TextField'

import { TextField as FormikTextField } from '../../ui/TextField/TextField'
import Modal from '../../ui/Modal/Modal'
import { AutocompleteField } from '../../ui/Autocomplete/Autocomplete'
import { messageType } from '../../../../shared/static'
import { User } from '@zbayapp/nectar/lib/sagas/users/users.slice'

const useStyles = makeStyles((theme) => ({
  root: {
    padding: 32,
    height: '100%',
    width: '100%'
  },
  fullContainer: {
    width: '100%',
    height: '100%'
  },
  gutter: {
    marginBottom: theme.spacing(4),
    marginTop: 0
  },
  button: {
    marginTop: 24,
    width: 165,
    backgroundColor: theme.palette.colors.zbayBlue,
    color: theme.palette.colors.white,
    '&:hover': {
      backgroundColor: theme.palette.colors.zbayBlue
    }
  },
  title: {
    marginBottom: 24
  }
}))

interface NewMessageModalProps {
  open: boolean
  handleClose: () => void
  sendMessage: (payload) => void
  users: User
}

export const NewMessageModal: React.FC<NewMessageModalProps> = ({
  open,
  handleClose,
  sendMessage,
  users,
}) => {
  const classes = useStyles({})
  const usersArray = Array.from(Object.values(users))
  return (
    <Modal open={open} handleClose={handleClose} title='' fullPage>
      <Grid className={classes.root}>
        <Formik
          initialValues={{
            recipient: '',
            message: '',
            user: {

            }
          }}
          onSubmit={(values, { resetForm }) => {
            const isAddressValid = /^t1[a-zA-Z0-9]{33}$|^ztestsapling1[a-z0-9]{75}$|^zs1[a-z0-9]{75}$/.test(
              values.recipient
            )
            const targetUser = usersArray.find(
              user => user.nickname === values.recipient || user.address === values.recipient
            )
            let payload
            if (targetUser) {
              payload = {
                message: values.message || '',
                spent: 0.0,
                type: messageType.BASIC,
                receiver: {
                  replyTo: targetUser.address,
                  username: targetUser.nickname
                }
              }
              sendMessage(payload)
              handleClose()
              resetForm()
              return
            }
            if (isAddressValid) {
              payload = {
                message: values.message || '',
                spent: 0.0,
                type: messageType.BASIC,
                receiver: {
                  replyTo: values.recipient,
                  username: values.recipient.substring(0, 15)
                }
              }
              sendMessage(payload)
              handleClose()
              resetForm()
              return
            }
          }}
        >
          {({ values, setFieldValue }) => (
            <Form className={classes.fullContainer}>
              <Grid
                container
                justify='flex-start'
                direction='column'
                className={classes.fullContainer}
              >
                <Typography variant='h3' className={classes.title}>
                  New message
                </Typography>
                <Typography variant='body2'>Recipient</Typography>
                <AutocompleteField
                  freeSolo
                  name={'recipient'}
                  inputValue={values.recipient || ''}
                  options={usersArray.map(option => option.nickname)}
                  value={values.recipient}
                  onChange={(v: string) => setFieldValue('recipient', v)}
                  onInputChange={(v: string) => {
                    setFieldValue('recipient', v)
                  }}
                  renderInput={params => (
                    <TextField
                      {...params}
                      className={classes.gutter}
                      variant='outlined'
                      multiline
                      maxRows={7}
                      placeholder={'Enter Zcash address or Zbay username'}
                      margin='normal'
                      fullWidth
                    />
                  )}
                />
                <Typography variant='body2'>Message text</Typography>
                <FormikTextField
                  name='message'
                  placeholder={'Enter message (optional)'}
                  InputProps={{}}
                  disabled={false}
                />
                <Button
                  className={classes.button}
                  variant='contained'
                  color='primary'
                  size='large'
                  type='submit'
                >
                  Send message
                </Button>
              </Grid>
            </Form>
          )}
        </Formik>
      </Grid>
    </Modal>
  )
}

export default NewMessageModal
