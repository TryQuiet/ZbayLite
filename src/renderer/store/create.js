import { applyMiddleware, createStore, compose } from 'redux'
import thunk from 'redux-thunk'
import promise from 'redux-promise-middleware'
import createDebounce from 'redux-debounced'

import reducers from './reducers'
import { errorsMiddleware } from './middlewares'

const composer = (enhancers) => {
  if (window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ &&
      process.env.NODE_ENV === 'development') {
    return window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__(enhancers)
  }
  return compose(enhancers)
}

export default ({
  initialState
} = {}) => createStore(
  reducers,
  initialState,
  composer(applyMiddleware(...[errorsMiddleware, createDebounce(), thunk, promise()]))
)
