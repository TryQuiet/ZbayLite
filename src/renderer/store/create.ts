import { applyMiddleware, createStore } from 'redux'
import thunk from 'redux-thunk'
import promise from 'redux-promise-middleware'
import createDebounce from 'redux-debounced'
import createSagaMiddleware from 'redux-saga'
import rootSaga from '../sagas/index.saga'
import { composeWithDevTools } from 'redux-devtools-extension'

import reducers from './reducers'
import { errorsMiddleware } from './middlewares'

const sagaMiddleware = createSagaMiddleware()

export default (initialState = {}) => {
  const store = createStore(
    reducers,
    initialState,
    composeWithDevTools(
      applyMiddleware(...[errorsMiddleware, createDebounce(), sagaMiddleware, thunk, promise()])
    )
  )
  sagaMiddleware.run(rootSaga)
  return store
}
