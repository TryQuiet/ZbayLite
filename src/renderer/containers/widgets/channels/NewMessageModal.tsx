import React from 'react'

import { useModal } from '../../hooks'
import NewMessageModal from '../../../components/widgets/channels/NewMessageModal'

import { ModalName } from '../../../sagas/modals/modals.types'

const NewMessage: React.FC = () => {
  const modal = useModal(ModalName.newMessageSeparate)
  return <NewMessageModal {...modal} users={{}} sendMessage={() => null} />
}
export default NewMessage
