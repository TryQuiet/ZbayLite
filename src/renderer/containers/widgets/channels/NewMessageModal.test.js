/* eslint import/first: 0 */
import { mapStateToProps, mapDispatchToProps } from './NewMessageModal'

import create from '../../../store/create'

describe('NewMessageModal', () => {
  let store = null
  beforeEach(() => {
    jest.clearAllMocks()
    store = create({
      users: {
        testuser1: 'user 1 data',
        testuser2: 'user 2 data'
      }
    })
  })

  it('will receive right props', () => {
    const props = mapStateToProps(store.getState())
    expect(props).toMatchSnapshot()
  })

  it('will receive right actions', () => {
    const actions = mapDispatchToProps(x => x)
    expect(actions).toMatchSnapshot()
  })
})
