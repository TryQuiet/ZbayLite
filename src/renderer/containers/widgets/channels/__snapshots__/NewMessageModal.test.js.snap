// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewMessageModal will receive right actions 1`] = `
Object {
  "showNotification": [Function],
}
`;

exports[`NewMessageModal will receive right props 1`] = `
Object {
  "users": Array [],
}
`;
